### ISSUE TITLE
Inside Emacs Intero REPL buffer quickBatch spams REPL with ^H characters

### ISSUE THREAD

**Issue by AlainODea** [commercialhaskell/intero#194]
If I load **QuickBatchReplSpamExample.hs** into the Intero REPL within Emacs and run `main`... (continued after code listing)

**QuickBatchReplSpamExample.hs**:

```haskell
module QuickBatchReplSpamExample where

import Control.Applicative
import Test.QuickCheck hiding (Success)
import Test.QuickCheck.Checkers
import Test.QuickCheck.Classes

newtype Identity a = Identity a deriving (Show, Eq)

instance Functor Identity where
  fmap f (Identity x) = Identity $ f x

instance Applicative Identity where
  pure = Identity
  (<*>) (Identity f) = fmap f

instance Arbitrary a => Arbitrary (Identity a) where
  arbitrary = Identity <$> arbitrary

instance Eq a => EqProp (Identity a) where (=-=) = eq

identityTrigger :: Identity (String, String, Int)
identityTrigger = undefined

main :: IO ()
main = do
  quickBatch $ applicative identityTrigger
```

I get output like this:
```plain
identity:     (0 tests)^H^H^H^H^H^H^H^H^H         ^H^H^H^H^H^H^H^H^H(1 test)^H^H^H^H^H^H^H^H         ^H^H^H^H^H^H^H^H(2 tests)^H^H^H^H^H^H^H^H^H         ^H^H^H^H^H^H^H^H^H(3 tests)^H^H^H^H^H^H^H^H^H         ^H^H^H^H^H^H^H^H^H(4 tests)^H^H^H^H^H^H^H^H^H         ^H^H^H^H^H^H^H^H^H(5 tests)^H^H^H^H^H^H^H^H^H         ^H^H^H^H^H^H^H^H^H(6 tests)^H^H^H^H^H^H^H^H^H         ^H^H^H^H^H^H^H^H^H(7 tests)^H^H^H^H^H^H^H^H^H         ^H^H^H^H^H^H^H^H^H(8 tests)^H^H^H^H^H^H^H^H^H
```

Output continues like that for the 500 tests. It also runs a lot slower than it does in Terminal or in eshell/shell buffers (seems to be lagged by Emacs rendering).

The **^H**s are rendered as if there is an encoding glitch. The caret moves from ^ to ^ and copying sections of the output from Emacs does not include these characters, but includes what appear to be an equivalent number of zero-width spaces.

This is reproducible in vanilla GNU Emacs 24.5 with nothing but this in .emacs and no pre-existing .emacs.d:

```elisp
;; If you don't have MELPA in your package archives:
(require 'package)
(add-to-list
  'package-archives
  '("melpa" . "http://melpa.org/packages/") t)
(package-initialize)
(package-refresh-contents)

;; Install Intero
(package-install 'intero)
(add-hook 'haskell-mode-hook 'intero-mode)
```

This is on Mac OS X and running **stack ghci --with-ghc intero --test** in a **shell** or **eshell** buffer works correctly (minus incremental feedback). In Terminal it works correctly and shows incremental progress.

My guess is that the Intero REPL buffer in Emacs may be representing itself to the Intero REPL child process as a terminal with more capabilities than it can actually provide (like treating **^H** as a backspace).

**Comment by chrisdone** [comment]
I thought that comint-mode handled this. Curious. Thanks for reporting.

**Comment by AlainODea** [comment]
My pleasure. Happy to test fixes or chase it down with some pointers to likely roots.

**Comment by chrisdone** [comment]
Does the same thing happen in `M-x shell` if you launch `stack ghci` and run the test in there?

**Comment by AlainODea** [comment]
If I use `M-x shell` and launch `stack ghci --test` and run the test in there it shows `+++ OK, passed 500 tests.` on each batch (identity, composition, etc.) without showing the intermediate `(N tests)` updates.

**Comment by chrisdone** [comment]
Good data.  Then there's something going on in the comint-derived mode, or shell-mode itself adds something extra to handle `^H`. We can look into that.

**Comment by AlainODea** [comment]
Thank you :)

**Comment by AlainODea** [comment]
Fixed by https://github.com/commercialhaskell/intero/commit/c2a8926811593bfcb51a9acddaa6a968227881ff from https://github.com/commercialhaskell/intero/pull/198 :)

Great work @angerman and @chrisdone 👍 

Thank you :)

**Comment by angerman** [comment]
@AlainODea Would you mind closing this issue, if it works for you now? :-)

**Comment by AlainODea** [comment]
Confirmed fixed in the MELPA released version of intero as of 20160804.346. Had to restart Emacs after package upgrade to get it to take effect. Thanks again :)
